### ISSUE TITLE
Feature Request: report bootstrap probabilities across hierarchy even when flat mapping is used

### ISSUE THREAD

**Issue by berl** [AllenInstitute/cell_type_mapper#27]
When flat mapping is used, the bootstrapping probability results at the finest resolution of the hierarchy accurately report the fraction of time a cell is mapped to the most-mapped category.  This same probability is currently reported across all levels of the hierarchy.  At other levels of the hierarchy, however, it's not accurately reflecting the fraction of bootstrap trials that matched the most-mapped category _at that level_. This would provide useful information about the stability of mapping results through the levels of the hierarchy. For example, in a reference with Subclass and Cluster levels, bootstrapping across clusters resulting in a `Cluster_bootstrapping_probability == 0.5` could stay entirely within the parent Subclass (`Subclass_bootstrapping_probability == 1.`) or the bootstrapping results could span multiple Subclasses (say, `Subclass_bootstrapping_probability == 0.6`). 

These hierarchical bootstrap results could help users understand how mapping quality and stability is distributed across cells and across a reference taxonomy.

**Comment by danielsf** [comment]
@berl

Thanks for your patience. I now have a branch that should implement this functionality for you, if you would like to test it out. The branch is here

https://github.com/AllenInstitute/cell_type_mapper/tree/correct/mapping/probabilities/250117

There was one subtlety that needed addressing. It is possible that, when doing flat mapping, the most probable subclass (for instance) is not consistent with the most probable cluster. Suppose that cluster A got 3 votes but clusters B, C, D, and E each got 1 vote. If clusters B, C, D, and E all belong to the same subclass which is not the subclass that cluster A belongs to, majority vote would naively assign the cell to cluster A and subclass BCDE, even though that makes no sense given the inheritance structure of the taxonomy. I decided to address this by forcing the mapper to assign the cell to a taxonomically-consistent set of cell types. In our cartoon example, the cell would get assigned to subclass A even though subclass BCDE technically got more votes. The JSON output (where runner up assignments are reported) will list subclass BCDE as the first runner up and should have enough data in it to show users that subclass BCDE actually got more votes that subclass A.

Do you agree with this solution?

Assuming everything checks out with this branch, the path forward will be to merge it into `rc/v1.5.0`. I will then need to update documentation (since this changes the interpretation of some of the quality metrics reported in the case of flat mapping) and decide when to release version 1.5.0

Thanks for this suggestion. I do think it is an improvement over the previous behavior. It was good to have the motivation to actually do this work.

**Comment by berl** [comment]
thanks for this, @danielsf!  Interesting point about the potential confusion with the hierarchy.   

pinging @scseeman and @omiethescientist for their thoughts on this

**Comment by scseeman** [comment]
Huh, this is tricky. On the one hand I'd agree with your solution @danielsf, but on another I could argue that the mapping seems more "confident" that the cell in your example is part of Subclass BCDE, it just can't place it accurately beyond that. I wonder, generally how appropriate it is to feed these probabilities back up the hierarchy in flat mapping when that's not what the mapping is doing in the first place. Maybe a safer alternative is to remove bootstrapping probabilities at all from other levels? Is it possible to flat-map at different levels? Then one could more accurately assess the most likely grouping at the level they are interested in and see how the bootstrapping comes out.

Looping @memoryleine in as well

**Comment by danielsf** [comment]
I feel like this conversation is getting very close to re-inventing hierarchical mapping (wherein the mapper first assigns the cell to the class it is most confident in, then assigns the cell to a subclass under that class, etc.)

May I ask what the advantage is to using flat mapping in place of hierarchical mapping?

"Can you flat map to different levels?"

No. However, there is a `drop_level` config parameter which tells hierarchical mapping to skip over a level when doing the mapping. We typically use this in whole mouse brain, since it was decided that we did not want to actually map to supertype, but just infer supertype from the cluster assignment.

I could make `drop_level` accept a list. Then, if you wanted to map directly to supertype, you could set

```
drop_level = ["CCN20230722_CLAS", "CCN20230722_SUBC"]
```

and the algorithm would skip over the class and subclass and start mapping directly to supertype (which I think would be effectively the same thing as flat mapping to the supertype level, for instance).

**Comment by scseeman** [comment]
Yeah, I don't think we want to re-invent hierarchical mapping. My understanding of hierarchical mapping is that is uses prescribed gene sets at each level for each grouping to do the mapping. This is fine when you have thousands of genes, it's less fine when you have a couple of hundred and now some (maybe all) of the genes in that prescribed set are missing and thus the mapping is horrible. Alternatively, if you just let the mapping use all of the limited number of genes, it can figure it out. Possibly there is a way to do hierarchical mapping but with all of the genes (assuming my understanding is correct)?

**Comment by danielsf** [comment]
Hierarchical mapping does use a lookup table that maps cell types to lists of markers good at discriminating between those types' children. When you run flat mapping, that table gets collapsed into one list (of about 6,500 genes for our whole mouse taxonomy) of genes that is used to map directly to the cluster level.

It would be easy to add a config parameter to the cell type mapper that would flatten the list of markers without flattening the taxonomy so that you were still doing hierarchical mapping, but you were using the same list of markers at every juncture.

It would also be easy to add a config parameter that told the mapper "I do not have a list of marker genes; just use all the genes in my unlabeled dataset."

The caveat being that if you used either of these flags on a dataset that had 10s of thousands of genes, the mapping would slow *way* down, but that is the user's choice.

Are either (both?) of these what you want @scseeman ?

**Comment by scseeman** [comment]
I think we definitely want the second thing ("...use all of the genes in my unlabeled dataset"), we struggle with that a lot in the current set up. The first thing also sounds interesting and I'd be keen to try it.

@berl apologies if I took this off-course, thoughts?

ps, @danielsf, maybe there could be a warning if someone mistakenly tried to use either of these features on a large, unlabeled dataset?

**Comment by danielsf** [comment]
I'll implement the proposed functionality in a different branch. I'll add a config param `collapse_markers`. If `collapse_markers = True` and a marker lookup table is specified, it will do the first thing I suggested (flatten the markers in the lookup table and use every marker at every layer in the hierarchy). If `collapse_markers = True` and a marker lookup is not specified, it will just use all of the genes in the unlabeled dataset. It will throw a warning if this results in it using >= 10,000 genes.

I would like to resolve the question of whether the current branch/feature gets merged.

My argument in favor of what I have done (force taxonomical consistency and report bootstrapping_probability) is that

a) there's no good way to flag "I assigned subclassBCDE, but that is not consistent with the fact that I also assigned clusterA." Users have to dig into the taxonomy and realize something doesn't make sense, and I'm not sure how they would know they need to do that.

b) noticing that there is a level in the taxonomy where bootstrapping_probability falls off a cliff would be a good indication that "the mapping is reliable up to this level, but is untrustworthy below that"

**Comment by berl** [comment]
@danielsf I actually like the idea of a) above (realizing you weren't exactly proposing it!).  It doesn't need to be at every level, just a single flag (per cell) indicating that there is an inconsistency in the results. if you're forcing the hierarchy onto the output (which I think is reasonable) you could compare at each level and if any are different, flag that cell.

**Comment by danielsf** [comment]
I just want to reiterate to make sure I understand what the proposal is.

In the case of flat mapping (with bootstrap_iterations > 1),
a) use majority vote to assign the cluster
b) force the higher levels in the taxonomy to be consistent with the cluster, even if that is inconsistent with "majority rules" voting
c) add a column "hierarchy_consistent" to the CSV that is True if (b) did not force any assignments to change and is False if (b) forced us to pick an assignment that was not actually the highest vote getter

The new column will not appear in the CSV in the cases of hierarchical mapping or bootstrap_iteration=1, since in these cases the mapping algorithm has already been forced to be taxonomically consistent.

?

I support this proposal. I was a little hesitant making the output of the web app more complicated, but we run correlation mapping with bootstrap_iterations = 1 in the web app, so this should not affect that experience.

**Comment by berl** [comment]
sounds great @danielsf

**Comment by danielsf** [comment]
Okay: I have added the new column (same branch: correct/mapping/probabilities/250117)

Now, if you run with
```
flatten = True
bootstrap_iteration > 1
```
a column `hierarchy_consistent` will appear in the CSV file. If `hierarchy_consistent == True`, it means that majority vote naturally produced type assignments that were consistent with the taxonomic hierarchy. If 'hierarchy_consistent == False`, it means that we had to force the cell's assignments to be consistent with the taxonomic hierarchy. The type assignments that actually run majority vote will be listed as `runners_up` in the JSON file (and will be identifiable because their `bootstrapping_probability` will be larger than the assigned `bootstrapping_probability`).

**Note:** In all cases, the cell type assignments reported in the CSV will be consistent with the taxonomic hierarchy.

Let me know when you feel comfortable with these changes, and I will merge them into the release candidate branch.

**Comment by danielsf** [comment]
I've merged this functionality in to `rc/v1.5.0`. For now, work off this branch if you want to exercise this functionality.

I'll leave this issue open until `rc/v1.5.0` gets merged into `main`. Go ahead and speak up here if something related to this needs to be addressed.

**Comment by danielsf** [comment]
I have merged this into `main`.
